core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/pythonworker/PythonProxyServer.scala

```diff
@@ -63,10 +66,20 @@ private class AmberProducer(
 
     val root = flightStream.getRoot
 
+    // send back ack with credits on ackStream
+    val bufferAllocator = new RootAllocator(8 * 1024)
+    try {
+      val arrowBuf: ArrowBuf = bufferAllocator.buffer(4 * 1024)
+      arrowBuf.writeLong(
+        31L
+      ) // TODO : replace with actual credit value
+      ackStream.onNext(PutResult.metadata(arrowBuf))
+      arrowBuf.close()
+    } finally if (bufferAllocator != null) bufferAllocator.close()
+
     // consume all data in the stream, it will store on the root vectors.
-    while (flightStream.next) {
-      ackStream.onNext(PutResult.metadata(flightStream.getLatestMetadata))
-    }
+    while (flightStream.next) {}
```

Comment:
I am confused with this logic. I think we should consume the stream before we send ack back?

Reply:
@Yicong-Huang I originally tried consuming the stream first, but it still hangs and the workflow doesn't finish executing just as I showed the last time.

I am wondering if the following scenario is happening when I try consuming the stream first:
1) The Python client starts to send data to Scala proxy via `writer.write_table` inside `proxy_client.py` line 66
2) The Python client then tries to read back the `credits` metadata the Scala proxy server sends back, but it immediately blocks to wait for this before it can finish sending data via `write_table` 
3) The Scala proxy server starts consuming the stream, but is also hangs because it must wait for the Python client to finish sending all data before sending back an ack
4) A deadlock is created

Does this seem like a reason why this could be happening?
Perhaps there could be other ways to have the Python client fully finish sending the data before trying to read the metadata?  I wasn't able to find any functions in the [py_arrow library](https://github.com/apache/arrow/blob/62829c5ca81e5ddaef2a609745aae6e6d08d1e6b/python/pyarrow/_flight.pyx) but perhaps there are other ways